sympy/vector/lamecoeff.py

```diff
@@ -0,0 +1,77 @@
+from sympy.core.basic import Basic
+from sympy import sqrt, sin, cos
+from sympy.core.function import Derivative
+
+
+class CoeffProvider(Basic):
+    """
+    It provides specific information of curvilinear coordinate
+    system to CoordSysCartesian.
+
+    Ideally, users should not instantiate this class.
+
+    """
+
+    def __new__(cls, system, curv_coord_name=None, transformation_equations=None):
+        obj = super(CoeffProvider, cls).__new__(cls, system)
+        obj.sys = system
+        obj._x, obj._y, obj._z = system.x, system.y, system.z
+        if curv_coord_name is not None:
+            class DefinedCoefficient:
```

Comment:
The transformation equations are a bit of a complicated thing, because they should affect the coordinate system itself and the way it relates to other coordinate system, it is something that should handled by the coordinate system class, not here. Let's postpone this part.

I don't see the need of using so many classes. I believe we need only one class definition in this file (that is, the class identifying the differential behavior of the operators, which is defined by the Lamé coefficients).